src/thirdparty/LightGlue/matcher.cpp

```diff
@@ -0,0 +1,603 @@
+                    desc0 = desc0.index_select(1, keep0);
+                    encoding0 = encoding0.index_select(-2, keep0);
+                    prune0.index_put_({torch::indexing::Slice(), ind0}, prune0.index({torch::indexing::Slice(), ind0}) + 1);
+                } else
+                {
+                    std::cout << "No points kept after pruning for desc0." << std::endl;
+                }
+            }
+
+            if (do_point_pruning && desc1.size(-2) > pruning_th)
+            {
+                auto scores1 = log_assignment_[i]->get_matchability(desc1);
+                auto prunemask1 = get_pruning_mask(token1, scores1, i);
+                if (prunemask1.dtype() != torch::kBool)
+                {
+                    prunemask1 = prunemask1.to(torch::kBool);
+                }
+
+                auto where_result = torch::where(prunemask1);
+                auto keep1 = where_result[1];
+                if (keep1.numel() > 0)
+                {
+                    ind1 = ind1.index_select(1, keep1);
+                    desc1 = desc1.index_select(1, keep1);
+                    encoding1 = encoding1.index_select(-2, keep1);
+                    prune1.index_put_({torch::indexing::Slice(), ind1}, prune1.index({torch::indexing::Slice(), ind1}) + 1);
+                } else
+                {
+                    std::cout << "No points kept after pruning for desc1." << std::endl;
+                }
+            }
+        }
+
+        // Handle empty descriptor case
+        if (desc0.size(1) == 0 || desc1.size(1) == 0)
+        {
+            auto m0 = torch::full({b, m}, -1, torch::TensorOptions().dtype(torch::kLong).device(device_));
+            auto m1 = torch::full({b, n}, -1, torch::TensorOptions().dtype(torch::kLong).device(device_));
+            auto mscores0 = torch::zeros({b, m}, device_);
+            auto mscores1 = torch::zeros({b, n}, device_);
+
+            if (!do_point_pruning)
+            {
+                prune0 = torch::ones_like(mscores0) * config_.n_layers;
+                prune1 = torch::ones_like(mscores1) * config_.n_layers;
+            }
+
+            torch::Dict<std::string, torch::Tensor> output;
+            output.insert("matches0", m0);
+            output.insert("matches1", m1);
+            output.insert("matching_scores0", mscores0);
+            output.insert("matching_scores1", mscores1);
+            output.insert("stop", torch::tensor(i + 1));
+            output.insert("prune0", prune0);
+            output.insert("prune1", prune1);
+
+            return output;
+        }
+
+        // Remove padding and compute assignment
+        desc0 = desc0.index({torch::indexing::Slice(), torch::indexing::Slice(0, m), torch::indexing::Slice()});
+        desc1 = desc1.index({torch::indexing::Slice(), torch::indexing::Slice(0, n), torch::indexing::Slice()});
+
+        auto scores = log_assignment_[std::min(i, config_.n_layers - 1)]->forward(desc0, desc1);
```

Comment:
@MrNeRF I believe here is a bug. I had to add the std::min() here to prevent segfault. Will be great for you to confirm that the fix is correct. Thx.

Reply:
Indeed, great catch! That's a stupid bug. The i variable belongs into the loop and should not be defined and declared outside. This way you can define a variable that gets i assigned the correct value within the for-loop which will be always std::min(i, config_.n_layers - 1).